### PR TITLE
Chart view for latency/jitter/packet loss

### PR DESCRIPTION
**Pull request by mateuszmidor**
KNTK-231
Based on https://github.com/kentik/sla_dashboard_webapp/pull/14

### CODE REVIEW

**Comment by mmac-m3a** [comment]
Please, rebase the pull request.

**Comment by mateuszmidor** [comment]
> Please, rebase the pull request.

Done.

**Review by lukaa12 — APPROVED** [review]

**Review by Fenthick — APPROVED** [review]

File: presentation/chart_view.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+from domain.model.mesh_results import HealthItem, MeshResults
+from domain.types import AgentID
+from presentation.health import Health
+
+
+class ChartView:
+    @staticmethod
+    def make_layout(from_agent, to_agent: AgentID, metric: Metric, results: MeshResults) -> html.Div:
+        # make chart title
+        agents = Agents(results)
```
Comment (mmac-m3a) [inline presentation/chart_view.py:22]: With my suggestion above, this would be:
```
from_alias = results.agents.get_alias(from_agent)
to_alias = results.agents.get_alias(to_agent)
```

File: presentation/chart_view.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        from_alias = agents.get_alias(from_agent)
+        to_alias = agents.get_alias(to_agent)
+        title = f"{metric.value}: {from_alias} -> {to_alias}"
+
+        # make chart data
+        health = Health(from_agent, to_agent, results)
+        xdata = [ChartView._utc_to_localtime(item.time) for item in health.items]
```
Comment (mmac-m3a) [inline presentation/chart_view.py:29]: `utc_to_localtime` should be helper function as opposed to a "private" static method of `ChartView`. It is not specific to `ChartView` and it may come handy at other places too.

File: presentation/health.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from typing import List
+
+from domain.model import HealthItem, MeshResults
+from domain.types import AgentID
+
+
+class Health:
```
Comment (mmac-m3a) [inline presentation/health.py:7]: I suggest to make the extraction of test results for particular agents and metric be part of the `MeshResults` class. The interface should be something like:

```
def filter(from_agent: AgentID, to_agent: AgentID, metric: Metric) -> List[datetime, float]:
   ...
```

The "health" terminology is misplaced here. "Health" (in the context of network monitoring) is a judgement based on data. What the "Health" class stores are actual observations.

File: infrastructure/data_access/http/synthetics_repo.py [hunk 4]
```diff
@@ -77,6 +78,20 @@ def transform_to_internal_mesh_columns(input_columns: List[V202101beta1MeshColum
...
         )
         columns.append(column)
     return columns
+
+
+def transform_to_internal_health_items(input_health: List[V202101beta1MeshMetrics]) -> List[HealthItem]:
+    health: List[HealthItem] = []
+    for h in input_health:
+        item = HealthItem(
+            jitter_microsec=int(h.jitter.value),
```
Comment (mmac-m3a) [inline infrastructure/data_access/http/synthetics_repo.py:91]: Let's do the scaling to milliseconds here as opposed to at every place we use the data. The value though should be `float`. We may end=up showing bunch of zeroes for jitter smaller than 1 ms.

```
jitter_millisec = h.jitter.value / 1000
...
```

File: presentation/chart_view.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    @staticmethod
+    def make_layout(from_agent, to_agent: AgentID, metric: Metric, results: MeshResults) -> html.Div:
+        # make chart title
+        agents = Agents(results)
+        from_alias = agents.get_alias(from_agent)
+        to_alias = agents.get_alias(to_agent)
+        title = f"{metric.value}: {from_alias} -> {to_alias}"
+
+        # make chart data
+        health = Health(from_agent, to_agent, results)
```
Comment (mmac-m3a) [inline presentation/chart_view.py:28]: Filtering of results to a particular agent pair and metric should be a method of the `MeshResults` class.

File: presentation/chart_view.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        xdata = [ChartView._utc_to_localtime(item.time) for item in health.items]
+        ydata = ChartView._get_ydata(metric, health)
```
Comment (mmac-m3a) [inline presentation/chart_view.py:30]: With all my suggested changes in place, data extraction and preparation would look like this:

```
series = results.filter(from_agent, to_agent, metric)
fig = go.Figure(data=[go.Scatter(x=[utc_to_localtime(e[0]) for e in series], y=[e[1] for e in series])])
```

File: presentation/chart_view.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        return f"/chart?from={from_agent}&to={to_agent}&metric={metric.value}"
+
+    @staticmethod
+    def decode_path(path: str) -> Tuple[AgentID, AgentID, Metric]:
+        path = unquote(path)
+        params = urlparse.parse_qs(urlparse.urlparse(path).query)
+        return params["from"][0], params["to"][0], Metric(params["metric"][0])
+
+    @staticmethod
+    def _utc_to_localtime(time: datetime) -> datetime:
```
Comment (mmac-m3a) [inline presentation/chart_view.py:61]: Simpler version:

```
def utc_to_localtime(time: datetime) -> datetime:
    return time.replace(tzinfo=timezone.utc).astimezone(tz=None)
```

File: presentation/error_404_view.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import dash_html_components as html
+
+
+class Error404View:
+    @staticmethod
+    def make_layout() -> html.Div:
```
Comment (mmac-m3a) [inline presentation/error_404_view.py:6]: Why not have a single `ErrorView` class and pass the error code as an argument to `make_layout`?

```
class ErrorView:
    ERROR_MESSAGE = {
         404: "404 Page not found",
         500: "500 Internal server error"
    }

    @classmethod
    def make_layout(cls, code: int) -> html.Div:
        return html.Div(
            children=[html.H1(children=cls.ERROR_MESSAGE[code],
                                           style={"textAlign": "center", "marginBottom": 50})],
        )
```

**Review by mmac-m3a — CHANGES_REQUESTED** [review]
I'd like you to consider my proposals now. I think they should be easy to implement and would simplify the code. Simplicity of the code is actually important, because we expect (at least some) customers to modify code for their own consumption.

File: main.py [hunk 9]
```diff
@@ -51,11 +80,44 @@ def get_production_server(self) -> flask.Flask:
     def run_development_server(self) -> None:
         self._app.run_server(debug=True)
 
-    def _make_layout(self) -> html.Div:
+    def _handle_path_change(self, pathname: str) -> html.Div:
+        try:
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply main.py:84]: Dash is said to be designed for single-page apps and unfortunately doesnt support @handler("/chart") def render_chart(...) style routing

File: main.py [hunk 10]
```diff
@@ -51,11 +80,44 @@ def get_production_server(self) -> flask.Flask:
...
+
+        agents = Agents(self._cached_repo.get_mesh_test_results())
+        from_agent_id = agents.get_id(from_agent_alias)
+        to_agent_id = agents.get_id(to_agent_alias)
+
+        return self._redirect_to_chart_view(from_agent_id, to_agent_id, self._current_metric)
+
+    def _make_matrix_layout(self) -> html.Div:
         mesh_test_results = self._cached_repo.get_mesh_test_results()
         results_timestamp = self._cached_repo.data_timestamp()
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply main.py:108]: timestamp will be modified in PR18

File: infrastructure/data_access/http/synthetics_repo.py (hunk 4 above)
Comment (mateuszmidor) [reply infrastructure/data_access/http/synthetics_repo.py:91]: Updating to milliseconds

File: presentation/chart_view.py (hunk 5 above)
Comment (mateuszmidor) [reply presentation/chart_view.py:28]: Not sure if data filtering and transformation should be part of the data structure itself; you can come up with different filters and transformers and this usually only bloats the class. I dont see profit here

File: presentation/chart_view.py (hunk 6 above)
Comment (mateuszmidor) [reply presentation/chart_view.py:30]: Such one-liners are not really readable, IMO

Comment (Fenthick) [reply presentation/chart_view.py:30]: I agree that such long lines are too complicated. I personally like inlining variables and then doing line breaks like this:

```
fig = go.Figure(data=[go.Scatter(
    x=[utc_to_localtime(e[0]) for e in series],
    y=[e[1] for e in series],
)])
```

This is closer to functional style, when reader sees first _what_ is being done (go.Figure()) instead of _how_ it is being done (how xdata and ydata is calculated)

**Comment by mateuszmidor** [comment]
Answered Review1 comments, updated code to most of them.

File: domain/model/mesh_results.py [hunk 11]
```diff
@@ -49,11 +49,30 @@ class MeshRow:
...
+            if row.agent_id == id:
+                return row.agent_alias
+        return f"[agent_id={id} not found]"
+
+
 class MeshResults:
     """Internal representation of Mesh Test results; independent of source data structure like http or grpc synthetics client"""
 
-    def __init__(self) -> None:
+    def __init__(self, timestamp: datetime = datetime.utcnow()) -> None:
```
Comment (Fenthick) [inline domain/model/mesh_results.py:72]:
```suggestion
    def __init__(self, timestamp: datetime = datetime.now(timezone.utc)) -> None:
```

It seems to be safer to have timezone info in the timestamp. Even if it works now, timestamp might have other usages in the future.

See: https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow

File: main.py [hunk 12]
```diff
@@ -1,30 +1,44 @@
...
+from presentation.error_500_view import Error500View
+from presentation.index_view import IndexView
+from presentation.matrix_view import MatrixView
 
 FORMAT = "[%(asctime)-15s] [%(process)d] [%(levelname)s]  %(message)s"
 logging.basicConfig(level=logging.INFO, format=FORMAT)
 logger = logging.getLogger(__name__)
 
 
+def get_auth_email_token() -> Tuple[str, str]:
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply main.py]: Done

**Review by Fenthick — APPROVED** [review]

File: domain/model/mesh_results.py (hunk 11 above)
Comment (Fenthick) [reply domain/model/mesh_results.py]: Done

File: presentation/matrix_view.py [hunk 13]
```diff
@@ -1,153 +1,200 @@
...
-        return f"<b>{cell.packet_loss_percent.value:.1f}%</b>"
 
+class MatrixView:
+    MATRIX = "matrix"
+    METRIC_SELECTOR = "metric-selector"
+
+    @staticmethod
+    def make_layout(mesh: MeshResults, metric: Metric) -> html.Div:
+        localtime_timestamp = utc_to_localtime(mesh.utc_timestamp)
+        timestamp = localtime_timestamp.strftime("%H:%M:%S")
```
Comment (mmac-m3a) [inline presentation/matrix_view.py:26]: The displayed  timestamp must include date. Imagine if a user loads the page, goes away for  a day, loads the same page in another window/tab after coming back, gets distracted and looks on the one loaded yesterday. She/He has no way to tell which of the two pages is actually current.

File: presentation/matrix_view.py [hunk 14]
```diff
@@ -1,153 +1,200 @@
...
 
-def make_hover_text(mesh: MeshResults, matrix: Matrix) -> List[List[str]]:
-    text = []
-    for row in reversed(mesh.rows):
-        text_col = []
-        for col in row.columns:
+    @staticmethod
+    def make_matrix_data(mesh: MeshResults, metric: Metric, config: Config) -> Dict:
+        matrix = Matrix(mesh)
+        data = MatrixView.make_data(mesh, metric, matrix, MatrixView.get_tresholds(metric, config))
```
Comment (mmac-m3a) [inline presentation/matrix_view.py:60]: Why you need to pass the output from `MatrixView.get_tresholds(metric, config)` to `MatrixView.make_data`? Why cannot `MatrixView.make_data` call the same internally?

Is there really any benefit of all methods of the `MatrixView` class being static? Why not have real instances and store `config` and `matrix` in instance variables.

At minimum, methods that need to call other methods of the same class should be @classmethods as opposed to static, so that the name of the class is available to the them. The net effect look like this:

```
@classmethod
def make_matrix_data(cls, mesh: MeshResults, metric: Metric, config: Config) -> Dict:
    matrix = Matrix(mesh)
    data = cls.make_data(mesh, metric, matrix, cls.get_tresholds(metric, config))
    ...
```

Also, I'd say that `make_view_data` would be a better name for this method.

File: presentation/matrix_view.py [hunk 15]
```diff
@@ -1,153 +1,200 @@
...
 
-            latency_ms = matrix.cells[row.agent_alias][col.agent_alias].latency_microsec.value * 1e-3
-            jitter_ms = matrix.cells[row.agent_alias][col.agent_alias].jitter_microsec.value * 1e-3
-            loss = matrix.cells[row.agent_alias][col.agent_alias].packet_loss_percent.value
-            text_col.append(
-                f"{row.agent_alias} -> {col.agent_alias} <br>Latency: {latency_ms:.2f} ms, <br>Jitter: {jitter_ms:.2f} ms, <br>Loss: {loss:.1f}%"
+        return {"data": data, "layout": layout}
+
+    @staticmethod
+    def get_tresholds(metric: Metric, config: Config) -> Thresholds:
```
Comment (mmac-m3a) [inline presentation/matrix_view.py:75]: typo: correct spelling is `get_threshold`.

File: presentation/matrix_view.py [hunk 16]
```diff
@@ -1,153 +1,200 @@
...
+        text = []
+        for row in reversed(mesh.rows):
+            text_col = []
+            for col in row.columns:
+
+                latency_ms = matrix.cells[row.agent_alias][col.agent_alias].latency_millisec.value
+                jitter_ms = matrix.cells[row.agent_alias][col.agent_alias].jitter_millisec.value
+                loss = matrix.cells[row.agent_alias][col.agent_alias].packet_loss_percent.value
+                text_col.append(
+                    f"{row.agent_alias} -> {col.agent_alias} <br>Latency: {latency_ms:.2f} ms, <br>Jitter: {jitter_ms:.2f} ms, <br>Loss: {loss:.1f}%"
```
Comment (mmac-m3a) [inline presentation/matrix_view.py:166]: Can the `get_text` method be used here too?

**Review by mmac-m3a — APPROVED** [review]
As discussed at stand-up, the code can be merged, but comments must be addressed in the next pull request.

File: main.py [hunk 17]
```diff
@@ -51,11 +71,43 @@ def get_production_server(self) -> flask.Flask:
     def run_development_server(self) -> None:
         self._app.run_server(debug=True)
 
-    def _make_layout(self) -> html.Div:
+    def _get_page_content(self, pathname: str) -> html.Div:
+        try:
+            if pathname == "/":
+                return self._make_matrix_layout()
+            elif pathname.startswith("/chart"):
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply main.py:78]: This specific example won't work for us as chart view takes parameters:  /chart?from=343&to=565.
I will update the code to prevent "charting" case.
Agree that implementing real path routing is overkill at this stage and probably won't be necessary.

File: presentation/matrix_view.py (hunk 14 above)
Comment (mateuszmidor) [reply presentation/matrix_view.py:60]: This class requires a major refactor, will get this done.

File: main.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply main.py:75]: We can always contribute to the project https://github.com/plotly/dash/pulls :)

File: presentation/health.py (hunk 3 above)
Comment (mateuszmidor) [reply presentation/health.py]: After discussion, moving filtering functionality into MeshResults

File: presentation/chart_view.py (hunk 5 above)
Comment (mateuszmidor) [reply presentation/chart_view.py]: After discussion, moving filtering to MeshResults

File: presentation/chart_view.py (hunk 6 above)
Comment (mateuszmidor) [reply presentation/chart_view.py:28]: The code was not intended to follow functional style; my intention here is to make it easier to read. And to breakpoint, when necessary

Comment (Fenthick) [reply presentation/chart_view.py:28]: Ok, that was just my opinion on what I find the most readable.

File: main.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (mmac-m3a) [reply main.py:78]: I've assumed that the `pathname` contains only URL path without query parameters. If it includes parameters we can do it this way:
```
components = pathname.split("?")[0].split("/')
if len(components) == 0:
    return self._make_matrix_layout()
if components[1] == "chart":
    return self._make_chart_layout(pathname)
return HTTPErrorView.make_layout(404)
```

File: main.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mmac-m3a) [reply main.py:75]: Sure, or we can use a different framework :).

File: main.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply main.py:78]: I will wrap this eventually into separate tool-function
